File: programcreek/js/01-18-valid-palindrome.js


function isValid(text) {
    var text = String(text).trim().toLowerCase();

    if (text == '') return false;

    var left = 0;
    var right = text.length - 1;

    while (left < right) {
        var leftChar = text.charAt(left);
        var rightChar = text.charAt(right);

        if (!isGoodChar(leftChar)) {
            left ++;
        }
        else if (!isGoodChar(rightChar)) {
            right --;
        }
        else {
            if (leftChar != rightChar) {
                return false;
            }
            else {
                left ++;
                right -- ;
            }
        }
    }

    return true;

}

function isGoodChar(a) {
    return /^[a-z0-9]$/.test(a);
}


console.log(isValid('Red rum, sir, is murder')) ;
console.log(isValid('Red rum, sir, is abc')) ;

File: programcreek/python/01/01-18-valid-palindrome.py
import re

def isValid(text) :
    text = text.strip().lower()

    if text == '':
        return False

    left = 0
    right = len(text) - 1

    while left < right:
        leftChar = text[left]
        rightChar = text[right]

        if not isGoodChar(leftChar):
            left += 1
        elif not isGoodChar(rightChar):
            right -= 1
        else:
            if leftChar != rightChar:
                return False
            else:
                left += 1
                right -= 1

    return True

def isGoodChar(a):
    return re.search(r'^[a-z0-9]$', a) is not None


print isValid('Red rum, sir, is murder')